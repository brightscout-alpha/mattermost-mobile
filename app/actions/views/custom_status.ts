// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {Client4} from '@client/rest';
import {logError} from '@mm-redux/actions/errors';
import {UserTypes} from '@mm-redux/action_types';
import {getCurrentUser} from '@mm-redux/selectors/entities/common';
import {ActionFunc, DispatchFunc, batchActions, GetStateFunc} from '@mm-redux/types/actions';
import {UserCustomStatus} from '@mm-redux/types/users';

export function setCustomStatus(customStatus: UserCustomStatus): ActionFunc {
    return async (dispatch: DispatchFunc, getState: GetStateFunc) => {
        const user = getCurrentUser(getState());
        const oldCustomStatus = user?.props?.customStatus;

        if (!user.props) {
            user.props = {};
        }

        user.props.customStatus = JSON.stringify(customStatus);
        dispatch({type: UserTypes.RECEIVED_ME, data: user});
        dispatch({type: UserTypes.SET_CUSTOM_STATUS_REQUEST});

        try {
            await Client4.updateCustomStatus(customStatus);
        } catch (error) {
            user.props.customStatus = oldCustomStatus;
            dispatch(batchActions([
                {type: UserTypes.RECEIVED_ME, data: user},
                {type: UserTypes.SET_CUSTOM_STATUS_FAILURE, error},
                logError(error),
            ]));
            return {error};
        }

        dispatch({type: UserTypes.SET_CUSTOM_STATUS_SUCCESS});
        return {data: true};
    };
}

export function unsetCustomStatus(): ActionFunc {
    return async (dispatch: DispatchFunc) => {
        dispatch({type: UserTypes.UNSET_CUSTOM_STATUS_REQUEST});

        try {
            await Client4.unsetCustomStatus();
        } catch (error) {
            dispatch(batchActions([
                {type: UserTypes.UNSET_CUSTOM_STATUS_FAILURE, error},
                logError(error),
            ]));
            return {error};
        }

        dispatch({type: UserTypes.UNSET_CUSTOM_STATUS_SUCCESS});
        return {data: true};
    };
}

export function removeRecentCustomStatus(customStatus: UserCustomStatus): ActionFunc {
    return async (dispatch: DispatchFunc) => {
        try {
            await Client4.removeRecentCustomStatus(customStatus);
        } catch (error) {
            dispatch(logError(error));
            return {error};
        }

        return {data: true};
    };
}

export default {
    setCustomStatus,
    unsetCustomStatus,
    removeRecentCustomStatus,
};
