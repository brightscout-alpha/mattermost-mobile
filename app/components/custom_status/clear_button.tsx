// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import React from 'react';
import CompassIcon from '@components/compass_icon';
import {Theme} from '@mm-redux/types/preferences';

import {changeOpacity, makeStyleSheetFromTheme} from '@utils/theme';
import {preventDoubleTap} from '@utils/tap';
import {View} from 'react-native';

interface Props {
    handlePress: () => void;
    size?: number;
    theme: Theme;
    testID?: string;
}

const ClearButton = (props: Props) => {
    const {handlePress, size, theme, testID} = props;
    const style = getStyleSheet(theme);

    return (
        <View
            style={style.container}
            testID={testID}
        >
            <CompassIcon
                onPress={preventDoubleTap(handlePress)}
                name='close'
                size={size}
                color={theme.centerChannelBg}
                style={style.button}
            />
        </View>
    );
};

ClearButton.defaultProps = {
    size: 20,
};

export default ClearButton;

const getStyleSheet = makeStyleSheetFromTheme((theme: Theme) => {
    return {
        container: {
            height: 40,
            width: 40,
            flex: 1,
            alignItems: 'center',
            justifyContent: 'center',
        },
        button: {
            backgroundColor: changeOpacity(theme.centerChannelColor, 0.52),
            borderRadius: 1000,
            overflow: 'hidden',
        },
    };
});
