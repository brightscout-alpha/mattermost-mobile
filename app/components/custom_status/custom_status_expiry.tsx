// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import React from 'react';

import moment from 'moment-timezone';

import FormattedDate from '@components/formatted_date';
import FormattedTime from '@components/formatted_time';
import {getCurrentMomentForTimezone} from '@utils/timezone';
import {GlobalState} from '@mm-redux/types/store';
import {getBool} from '@mm-redux/selectors/entities/preferences';
import Preferences from '@mm-redux/constants/preferences';
import {useSelector} from 'react-redux';
import {Text, TextStyle} from 'react-native';
import FormattedText from '@components/formatted_text';
import {Theme} from '@mm-redux/types/preferences';
import {makeStyleSheetFromTheme} from '@utils/theme';
import {getCurrentUserTimezone} from '@mm-redux/selectors/entities/timezone';
type Props = {
    theme: Theme;
    time: Date;
    styleProp?: TextStyle;
    testID?: string;
    showPrefix?: boolean;
    withinBrackets?: boolean;
}

const CustomStatusExpiry = (props: Props) => {
    const {time, theme, styleProp, showPrefix, withinBrackets} = props;
    const timezone = useSelector(getCurrentUserTimezone);
    const styles = createStyleSheet(theme);
    const militaryTime = useSelector((state: GlobalState) => getBool(state, Preferences.CATEGORY_DISPLAY_SETTINGS, 'use_military_time'));
    const currentMomentTime = getCurrentMomentForTimezone(timezone);

    const expiryMomentTime = timezone ? moment(time).tz(timezone) : moment(time);

    const plusSixDaysEndTime = currentMomentTime.clone().add(6, 'days').endOf('day');
    const tomorrowEndTime = currentMomentTime.clone().add(1, 'day').endOf('day');
    const todayEndTime = currentMomentTime.clone().endOf('day');

    let useTime = true;
    let useDay = false;
    let isTomorrow = false;
    let isToday = false;
    let useDate = false;
    let format = '';

    if (expiryMomentTime.isSame(todayEndTime)) {
        isToday = true;
    }
    if (expiryMomentTime.isAfter(todayEndTime) && expiryMomentTime.isSameOrBefore(tomorrowEndTime)) {
        isTomorrow = true;
    }
    if (expiryMomentTime.isSame(todayEndTime) || expiryMomentTime.isAfter(tomorrowEndTime)) {
        useTime = false;
    }
    if (expiryMomentTime.isBetween(tomorrowEndTime, plusSixDaysEndTime)) {
        useDay = true;
    }

    const isCurrentYear = currentMomentTime.get('y') === expiryMomentTime.get('y');

    useDate = !(isToday || useTime || useDay || !isCurrentYear);

    if (useDay) {
        format = 'dddd';
    } else if (useDate) {
        format = 'MMM DD';
    } else if (!isCurrentYear) {
        format = 'MMM DD, YYYY';
    }

    const showDayorDate = (useDay || useDate || !isCurrentYear) ? (
        <FormattedDate
            format={format}
            timezone={timezone}
            value={expiryMomentTime}
        />
    ) : null;

    const showTime = useTime ? (
        <FormattedTime
            hour12={!militaryTime}
            timezone={timezone}
            value={expiryMomentTime}
        />
    ) : null;

    const showTomorrow = isTomorrow ? (
        <Text>
            <FormattedText
                id='custom_status.expiry_time.tomorrow'
                defaultMessage='Tomorrow'
            />{' at '}
        </Text>
    ) : null;

    const showToday = isToday ? (
        <FormattedText
            id='custom_status.expiry_time.today'
            defaultMessage='Today'
        />
    ) : null;

    const prefix = showPrefix ? (
        <Text>
            <FormattedText
                id='custom_status.expiry.until'
                defaultMessage='Until'
            />{' '}
        </Text>
    ) : null;

    return (
        <Text
            testID={props.testID}
            style={styleProp || styles.text}
        >
            {withinBrackets ? '(' : null}
            {prefix}
            {showToday}
            {showTomorrow}
            {showTime}
            {showDayorDate}
            {withinBrackets ? ')' : null}
        </Text>
    );
};

export default CustomStatusExpiry;

const createStyleSheet = makeStyleSheetFromTheme((theme: Theme) => {
    return {
        text: {
            fontSize: 15,
            color: theme.centerChannelColor,
        },
    };
});
