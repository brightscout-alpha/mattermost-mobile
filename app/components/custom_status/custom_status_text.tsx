// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import React from 'react';
import {changeOpacity, makeStyleSheetFromTheme} from 'app/utils/theme';
import type {Theme} from '@mm-redux/types/preferences';
import {Text, TextStyle} from 'react-native';

interface ComponentProps {
    text: string;
    theme: Theme;
    textStyle?: TextStyle
}

const CustomStatusText = ({text, theme, textStyle}: ComponentProps) => (
    <Text style={[getStyleSheet(theme).label, textStyle]}>
        {text}
    </Text>
);

export default CustomStatusText;

const getStyleSheet = makeStyleSheetFromTheme((theme: Theme) => {
    return {
        label: {
            color: changeOpacity(theme.centerChannelColor, 0.5),
            flex: 1,
            fontSize: 17,
            textAlignVertical: 'center',
            includeFontPadding: false,
        },
    };
});
