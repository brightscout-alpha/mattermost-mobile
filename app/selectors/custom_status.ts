// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {GlobalState} from '@mm-redux/types/store';
import {CustomStatusDuration, UserCustomStatus} from '@mm-redux/types/users';

import {createSelector} from 'reselect';
import moment from 'moment-timezone';

import {Preferences} from '@mm-redux/constants';
import {getConfig} from '@mm-redux/selectors/entities/general';
import {get} from '@mm-redux/selectors/entities/preferences';
import {getCurrentUser, getUser} from '@mm-redux/selectors/entities/users';
import {getCurrentUserTimezone} from '@mm-redux/selectors/entities/timezone';
import {isMinimumServerVersion} from '@mm-redux/utils/helpers';

import {getCurrentMomentForTimezone} from '@utils/timezone';

export function getCustomStatus(state: GlobalState, userID?: string): UserCustomStatus | undefined {
    const user = userID ? getUser(state, userID) : getCurrentUser(state);
    const userProps = user?.props || {};
    const customStatus = userProps.customStatus ? JSON.parse(userProps.customStatus) : undefined;
    const timezone = getCurrentUserTimezone(state);
    const expiryTime = moment(customStatus?.expires_at);
    const currentTime = getCurrentMomentForTimezone(timezone);
    return (customStatus?.duration === CustomStatusDuration.DONT_CLEAR || currentTime < expiryTime) ? customStatus : undefined;
}

export const getRecentCustomStatuses = createSelector(
    (state: GlobalState) => get(state, Preferences.CATEGORY_CUSTOM_STATUS, Preferences.NAME_RECENT_CUSTOM_STATUSES),
    (value) => {
        try {
            return value ? JSON.parse(value) : [];
        } catch (error) {
            return [];
        }
    },
);

export function isCustomStatusEnabled(state: GlobalState) {
    const config = getConfig(state);
    const serverVersion = state.entities.general.serverVersion;
    return config && config.EnableCustomUserStatuses === 'true';
}
