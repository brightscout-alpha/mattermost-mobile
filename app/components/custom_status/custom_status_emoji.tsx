// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import React from 'react';
import {makeGetCustomStatus, isCustomStatusEnabled} from '@selectors/custom_status';
import {useSelector} from 'react-redux';
import {GlobalState} from '@mm-redux/types/store';
import Emoji from '@components/emoji';
import {Text, TextStyle} from 'react-native';

interface ComponentProps {
    emojiSize?: number;
    userID?: string;
    style?: TextStyle;
    testID?: string;
}

const CustomStatusEmoji = (props: ComponentProps) => {
    const getCustomStatus = makeGetCustomStatus();
    const {emojiSize, userID, style, testID} = props;
    const customStatusEnabled = useSelector(isCustomStatusEnabled);
    const customStatus = useSelector((state: GlobalState) => {
        return getCustomStatus(state, userID);
    });
    const emojiExists = customStatusEnabled && customStatus && customStatus.emoji;
    if (!emojiExists) {
        return null;
    }

    const testIdPrefix = testID ? `${testID}.` : '';
    return (
        <Text
            style={style}
            testID={`${testIdPrefix}custom_status_emoji.${customStatus.emoji}`}
        >
            <Emoji
                size={emojiSize}
                emojiName={customStatus.emoji}
            />
        </Text>
    );
};

CustomStatusEmoji.defaultProps = {
    emojiSize: 16,
};

export default CustomStatusEmoji;
