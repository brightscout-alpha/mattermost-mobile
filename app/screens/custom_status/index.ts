// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {connect} from 'react-redux';
import {bindActionCreators, Dispatch} from 'redux';

import {getTheme} from '@mm-redux/selectors/entities/preferences';
import {GlobalState} from '@mm-redux/types/store';
import {getRecentCustomStatuses, makeGetCustomStatus} from '@selectors/custom_status';
import {setCustomStatus, unsetCustomStatus, removeRecentCustomStatus} from '@actions/views/custom_status';

import CustomStatusModal from '@screens/custom_status/custom_status_modal';
import {GenericAction} from '@mm-redux/types/actions';
import {isLandscape} from '@selectors/device';
import {getUserTimezone} from '@mm-redux/selectors/entities/timezone';
import {getCurrentUserId} from '@mm-redux/selectors/entities/users';

function mapStateToProps(state: GlobalState) {
    const getCustomStatus = makeGetCustomStatus();
    const customStatus = getCustomStatus(state) || {};
    const recentCustomStatuses = getRecentCustomStatuses(state);
    const theme = getTheme(state);
    const currentUserId = getCurrentUserId(state);
    const userTimezone = getUserTimezone(state, currentUserId);

    return {
        userTimezone,
        customStatus,
        recentCustomStatuses,
        theme,
        isLandscape: isLandscape(state),
    };
}

function mapDispatchToProps(dispatch: Dispatch<GenericAction>) {
    return {
        actions: bindActionCreators({
            setCustomStatus,
            unsetCustomStatus,
            removeRecentCustomStatus,
        }, dispatch),
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(CustomStatusModal);
