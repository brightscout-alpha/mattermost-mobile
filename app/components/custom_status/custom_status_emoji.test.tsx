// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {shallow} from 'enzyme';
import React from 'react';

import configureStore from 'redux-mock-store';
import {Provider} from 'react-redux';

import * as CustomStatusSelectors from '@selectors/custom_status';

import CustomStatusEmoji from '@components/custom_status/custom_status_emoji';
import Emoji from '@components/emoji';

jest.mock('@selectors/custom_status');

describe('components/custom_status/custom_status_emoji', () => {
    const mockStore = configureStore();
    const store = mockStore({});

    const getCustomStatus = () => {
        return {
            emoji: 'calendar',
            text: 'In a meeting',
        };
    };
    (CustomStatusSelectors.makeGetCustomStatus as jest.Mock).mockReturnValue(getCustomStatus);
    (CustomStatusSelectors.isCustomStatusEnabled as jest.Mock).mockReturnValue(true);
    it('should match snapshot', () => {
        const wrapper = shallow(
            <Provider store={store}>
                <CustomStatusEmoji/>
            </Provider>,
        );
        expect(wrapper.getElement()).toMatchSnapshot();
    });

    it('should match snapshot with props', () => {
        const wrapper = shallow(
            <Provider store={store}>
                <CustomStatusEmoji
                    emojiSize={34}
                />
            </Provider>,
        );

        expect(wrapper.getElement()).toMatchSnapshot();
    });

    it('should not render when EnableCustomStatus in config is false', () => {
        (CustomStatusSelectors.isCustomStatusEnabled as jest.Mock).mockReturnValue(false);
        const wrapper = shallow(
            <Provider store={store}>
                <CustomStatusEmoji/>
            </Provider>,
        );

        expect(wrapper.containsMatchingElement(<Emoji/>)).toBeFalsy();
    });

    it('should not render when getCustomStatus returns null', () => {
        (CustomStatusSelectors.isCustomStatusEnabled as jest.Mock).mockReturnValue(true);
        (CustomStatusSelectors.makeGetCustomStatus as jest.Mock).mockReturnValue(() => null);
        const wrapper = shallow(
            <Provider store={store}>
                <CustomStatusEmoji/>
            </Provider>,
        );

        expect(wrapper.containsMatchingElement(<Emoji/>)).toBeFalsy();
    });
});
