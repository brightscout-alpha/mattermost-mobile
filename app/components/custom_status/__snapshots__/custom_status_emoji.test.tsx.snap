// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/custom_status/custom_status_emoji should match snapshot 1`] = `
<Context.Provider
  value={
    Object {
      "store": Object {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "@@observable": [Function],
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <CustomStatusEmoji
    emojiSize={16}
  />
</Context.Provider>
`;

exports[`components/custom_status/custom_status_emoji should match snapshot with props 1`] = `
<Context.Provider
  value={
    Object {
      "store": Object {
        "@@observable": [Function],
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "@@observable": [Function],
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <CustomStatusEmoji
    emojiSize={34}
  />
</Context.Provider>
`;
